docs(v2): `challenger` (#693)
Layr-Labs/eigensdk-go: challenger/README.md

   ```

2. **Challenger Configuration**: Create a `challenger.Config` struct with the following fields:
   - `EthWsUrl`: The URL of the Ethereum websocket
   - `Logger`: The logger
   - `TaskManagerAbi`: The ABI of the task manager
   - `EthClient`: The Ethereum client

      ```go
        ethHttpUrl := "http://localhost:8545"
        ethHttpClient, _ := ethclient.Dial(ethHttpUrl)
      
        cfg := challenger.Config{
          EthWsUrl:       "ws://localhost:8545",
          Logger:         logger,
          TaskManagerAbi: taskManagerAbi,
          EthClient:      ethHttpClient,
        }
      ```

3. **Task Manager**: Create a `TaskManager` to interact with the user defined task manager contract.
   - This will be in charge of raising challenges.

      ```go
        ecdsaPrivateKey, _ := crypto.HexToECDSA(testutils.ANVIL_FIRST_PRIVATE_KEY)
        txMgr, _ := txmgr.NewSimpleTxManagerFromPrivateKey(logger, ethHttpClient, ecdsaPrivateKey)
  
        challengerRaiser, err := taskmanager.NewTaskManagerFromAbi[*big.Int, *big.Int](
          taskManagerAddress,
          taskManagerAbi,
          txMgr,
          ethHttpClient,
        )
      ```

> 4. **Task Verification Logic**: Define a function that computes the expected result for a task, which will be used to verify operator responses

Reviewer: Let's talk about directly implementing a verifier function here instead, like:

```go
func isValidSquare(taskIndex uint32, numberToSquare, numberSquared *big.Int) (bool, error) {
    expectedNumberSquared := big.NewInt(0).Exp(numberToSquare, big.NewInt(2), nil)
    return expectedNumberSquared.Cmp(numberSquared) == 0, nil
}
```

After that, we can mention the other approach briefly, just mentioning the `ResponseValidationFunctionFromResponseCalculator` function, and how it's used, without going through how one would create a response calculator (we can link to the operator documentation for this).

          numberSquared := big.NewInt(0).Exp(numberToSquare, big.NewInt(2), nil)
          return numberSquared, nil
        }
      ```

5. **Response Calculator**: To abstract your computation into the challenger, we provide a `ResponseCalculator` interface with a standar `functionResponseCalculator` struct. This struct implements the interface and a helper method for turning your function into `functionResponseCalculator`:
   - `NewFunctionResponseCalculator`: Create a response calculator from your computation function.

      ```go
        calculator := operator.NewFunctionResponseCalculator(square)
      ```

6. **Verifier**: Create a verifier from the response calculator.
   - This will be in charge of computing the response of a task and will use a user-defined function to compare the computed response with the operator's response.

      ```go
        // BigIntEqual receives two *big.Int and returns whether they are equal
        func BigIntEqual(a, b *big.Int) bool {
          return a.Cmp(b) == 0
        }

        validation := challengerprocessor.ResponseValidationFunctionFromResponseCalculator(calculator, BigIntEqual)
      ```

> 7. **Challenger Task Processor**: Create a [`ChallengerTaskProcessor`] interface implementation.

Reviewer: We should add a section specifying how one would go about for creating a ChallengerTaskProcessor implementation. Same for the other modules.

        }
      ```

> 2. **Task Verification Logic**: Define a function that computes the response for a task and compares it with the operator's response.

Reviewer: ```suggestion
2. **Task Verification Logic**: Define a function that verifies the response for a task, returning `true` for valid tasks.
```

    ```go
      func isValidSquare(taskIndex uint32, numberToSquare, numberSquared *big.Int) (bool, error) {
          expectedNumberSquared := big.NewInt(0).Exp(numberToSquare, big.NewInt(2), nil)
          return expectedNumberSquared.Cmp(numberSquared) == 0, nil
      }
    ```

    - There is another approach where you can use the logic from the operator to compute the task. To do this, you can wrap the logic into a `ResponseCalculator` implementation. Then you can use `ResponseValidationFunctionFromResponseCalculator`, which will be in charge of computing the response of a task and will use a user-defined function to compare the computed response with the operator's response.

> 3. **Provide a Challenger Processor**: Provide a struct implementing the `ChallengerProcessor` interface. This interface contains user-defined logic to process new tasks and task responses. We provide a standard `IndexingChallengerProcessor` implementation that can be used as is for most cases, you can create it using the `NewIndexingChallengerProcessor` constructor from `challengerprocessor` package:

Reviewer: ```suggestion
3. **Provide a Challenger Processor**: Provide a struct implementing the `ChallengerProcessor` interface. This interface contains user-defined logic to process new tasks and task responses. We provide a standard `IndexingChallengerProcessor` implementation that can be used as-is for most cases, you can create it using the `NewIndexingChallengerProcessor` constructor from `challengerprocessor` package:
```

    1. Instantiate an Ethereum client and a transaction manager for the `ChallengerProcessor`:

        ``` go
            ethHttpClient, err := ethclient.Dial("http://localhost:8545")

            ecdsaPrivateKey, err := crypto.HexToECDSA("2a871d0798f97d79848a013d4936a73bf4cc922c825d33c1cf7073dff6d409c6")
            txMgr, err := txmgr.NewSimpleTxManagerFromPrivateKey(logger, ethHttpClient, ecdsaPrivateKey)
        ```

    2. Get the `TaskManager` ABI from the contract binding:

        ``` go
            taskManagerAbi, err := cstaskmanager.ContractIncredibleSquaringTaskManagerMetaData.GetAbi()
        ```
    

>     3. Provide a struct that implements the `ChallengerRaiser` interface. That interface requires a method to raise an on-chain challenge when operator responses are incorrect. Here we provide a SDK implementation that satisfies the `ChallengerRaiser` interface, receiving the TaskManager address and ABI, a transaction manager and an Ethereum client:

Reviewer: The interface is called `ChallengeRaiser`

Author: Done [e7cd3e6](https://github.com/Layr-Labs/eigensdk-go/pull/693/commits/e7cd3e65c8fb38a68cd2d645d138daeb329f5a43)

- [Incredible Squaring](https://github.com/Layr-Labs/eigensdk-go/blob/v2-dev-1/examples/incredible-squaring/challenger/main.go)
- [Incredible Dot Product](https://github.com/Layr-Labs/eigensdk-go/blob/v2-dev-1/examples/incredible-dot-product/challenger/main.go)
- [Awesome Vault Service](https://github.com/Layr-Labs/eigensdk-go/blob/v2-dev-1/examples/awesome-vault-service/challenger/main.go)

## How to implement a custom Challenger Processor

> To implement a custom Challenger Processor, you must implement the `ChallengerProcessor` interface, which defines two methods:

Reviewer: Let's restructure this section:

```
## How to implement a custom Challenger Processor

To implement a custom Challenger Processor, you must implement the `ChallengerProcessor` interface:

<interface>

This interface has two methods, which we explain in the next sections. Refer to the [`IndexingChallengerProcessor`](https://github.com/Layr-Labs/eigensdk-go/blob/v2-dev-2/challenger/challenger-processor/challenger_task_processor.go) implementation for an example of how to implement a custom Challenger Processor.

### `ProcessNewTaskCreated`

Invoked when a new task is emitted by the contract. `ProcessNewTaskCreated` is commonly used to store the task (e.g., in a map, in a database, etc.), so that when a response arrives, you can retrieve the corresponding input.

### `ProcessTaskResponded`

Invoked when a task response is received. This method should verify the operator’s response and raise a challenge for invalid responses. Step by step, this method:

1. Retrieves the original task using the index.
2. Verifies the operator’s response against the task's input.
3. Raises a challenge through the `TaskManager` if the responses differ.
```

Author: Done [69bfe8f](https://github.com/Layr-Labs/eigensdk-go/pull/693/commits/69bfe8fb32a4fd973cb306b2f1880b9e3f8d88db)

```go
  type ChallengerProcessor[Input any, Output any] interface {
    ProcessNewTaskCreated(taskIndex uint32, task taskmanager.Task[Input]) error
    ProcessTaskResponded(taskIndex uint32, taskResponse taskmanager.TaskResponse[Output], taskResponseMetadata sdktypes.TaskResponseMetadata, nonSigningOperatorPubKeys []sdktypes.BN254G1Point) error
  }
```

- `ProcessNewTaskCreated`: Invoked when a new task is emitted by the contract. This method should store the task to be used later during response validation.

> - `ProcessTaskResponded`: Invoked when a task response is received. This method compares the operator’s response to the expected one and raises a challenge if they differ.

Reviewer: ```suggestion
- `ProcessTaskResponded`: Invoked when a task response is received. This method verifies the operator’s response and raises a challenge for invalid responses.
```
